src/libproc_macro/lib.rs

```diff
@@ -200,6 +201,78 @@ impl Span {
     pub fn call_site() -> Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
+
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<SourceFile> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt != SyntaxContext::empty() {
+            return None
+        }
+
+        let loc = __internal::lookup_char_pos(self.0.lo);
+
+        Some(SourceFile(loc.file))
+    }
+
+    /// Get the starting line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn start(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.lo);
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Get the ending line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn end(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.hi);
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Create a new span encompassing `self` and `other`.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn glue(&self, other: Span) -> Span {
+        Span(self.0.to(other.0))
+    }
+}
+
+/// A line-column pair representing the start or end of a `Span`.
+#[unstable(feature = "proc_macro", issue = "38356")]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct LineColumn {
+    /// The 1-indexed line in the source file on which the span starts or ends (inclusive).
+    line: usize,
+    /// The 0-indexed column (in UTF-8 characters) in the source file on which
+    /// the span starts or ends (inclusive).
+    column: usize
+}
+
+/// The name/path of the source file of a `Span`.
+#[unstable(feature = "proc_macro", issue = "38356")]
+#[derive(Clone)]
+pub struct SourceFile(Rc<syntax_pos::FileMap>);
+
+impl SourceFile {
+    /// Get the underlying file name/path.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn as_str(&self) -> &str {
```

Comment:
Does this return the path or the name of the file? If path: relative to the manifest-file? Relative to something else? Or absolute?

And if a path is returned, why not `std::path::Path` instead of `&str`? And then I'd also rename the method to `path()`.

Reply:
The underlying representation is a string but it is typically a path, yeah. I believe it's relative to the working directory for the current process.

Reply:
It's relative to the directory containing the file being compiled, i.e. `~/foo/src` for `rustc ~/foo/src/lib.rs`.
I think this method should be named `path` and should return an absolute path.

Reply:
@jseyfried Are you want it to be absolute? It'd be more difficult to unit-test.

Reply:
Perhaps the unit test could just check that a suffix of the absolute path is correct. I think it's worth having an absolute path so that the proc macro author doesn't have to worry about to which the path is relative.